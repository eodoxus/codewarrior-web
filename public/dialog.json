{
  "CrestfallenHome.CrestfallenMage": [
    "Hope...\nHope is forever lost...",
    "The Great Book of ECMAScript has been captured.\nAll I have left is this tattered page.",
    "It is said that a wizard in the north can release the book.\nIf only I could decipher the runes on this tattered page.\nPerhaps you can try?",
    "You've freed the Great Book!\nOh Hero, a great journey awaits you.\nIt is said the Great book can open many doors.\nGo east!"
  ]
}
