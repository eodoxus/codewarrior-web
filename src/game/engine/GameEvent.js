import Event from "../../lib/Event";

const listeners = [];
let inputQueue;

export default class GameEvent extends Event {
  static CANCEL = "cancel";
  static CLICK = "click";
  static CLICK_HERO = "clickHero";
  static CLICK_MENU = "clickMenu";
  static CLOSE_CURTAIN = "closeCurtain";
  static CLOSE_HERO_MENU = "closeHeroMenu";
  static CLOSE_TATTERED_PAGE = "closeTatteredPage";
  static CONFIRM = "confirm";
  static COLLISION = "collision";
  static DIALOG = "dialog";
  static DOORWAY = "doorway";
  static HIDE_BORDER = "hideBorder";
  static NPC_INTERACTION = "npcInteraction";
  static TALK = "talk";
  static TRANSITION = "transition";
  static OPEN_CURTAIN = "openCurtain";
  static OPEN_HERO_MENU = "openHeroMenu";
  static OPEN_TATTERED_PAGE = "openTatteredPage";
  static SHOW_BORDER = "showBorder";
  static STOP = "stop";

  static click(tile) {
    return new EventType(GameEvent.CLICK, tile);
  }

  static generic(eventName, data) {
    return new EventType(eventName, data);
  }

  static collision(entity) {
    return new EventType(GameEvent.COLLISION, entity);
  }

  static heroClick(hero) {
    return new EventType(GameEvent.CLICK_HERO, hero);
  }

  static menuClick(position) {
    return new EventType(GameEvent.CLICK_MENU, position);
  }

  static talk(entity) {
    return new EventType(GameEvent.TALK, entity);
  }

  static removeAllListeners() {
    listeners.forEach(listener => {
      listener.remove();
    });
  }

  static on(event, handler, enqueue = false) {
    const listener = super.on(event, handler);
    if (enqueue) {
      listeners.push(listener);
    }
    return listener;
  }

  static inputQueue(event) {
    if (!inputQueue) {
      inputQueue = new InputQueue();
    }
    return inputQueue;
  }
}

class EventType {
  type;
  data;
  constructor(type, data) {
    this.type = type;
    this.data = data;
  }

  getType() {
    return this.type;
  }

  getData() {
    return this.data;
  }
}

class InputQueue {
  queue;

  constructor() {
    this.queue = [];
  }

  add(event) {
    if (this.queue.length === 0) {
      this.queue.push(event);
      return;
    }
    if (this.queue[this.queue.length - 1].getType() !== event.getType()) {
      this.queue.push(event);
    }
  }

  getNext() {
    if (this.queue.length) {
      return this.queue.pop();
    }
  }
}
