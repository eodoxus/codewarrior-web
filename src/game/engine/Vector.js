export default class Vector {
  static add(v, addend) {
    return _vectorOperation(v, "add", addend);
  }

  static subtract(v, subtrahend) {
    return _vectorOperation(v, "subtract", subtrahend);
  }

  static multiply(v, factor) {
    return _vectorOperation(v, "multiply", factor);
  }

  static divide(v, quotient) {
    return _vectorOperation(v, "divide", quotient);
  }

  static normalize(v) {
    return _vectorOperation(v, "normalize");
  }

  static round(v) {
    return _vectorOperation(v, "round");
  }

  x;
  y;

  constructor(x, y) {
    this.x = x || 0;
    this.y = y || 0;
  }

  add(v) {
    if (typeof v === "number") {
      this.x += v;
      this.y += v;
    } else {
      this.x += v.x;
      this.y += v.y;
    }
    return this;
  }

  subtract(v) {
    if (typeof v === "number") {
      this.x -= v;
      this.y -= v;
    } else {
      this.x -= v.x;
      this.y -= v.y;
    }
    return this;
  }

  multiply(v) {
    if (typeof v === "number") {
      this.x *= v;
      this.y *= v;
    } else {
      this.x *= v.x;
      this.y *= v.y;
    }
    return this;
  }

  divide(v) {
    if (typeof v === "number") {
      this.x /= v;
      this.y /= v;
    } else {
      this.x /= v.x;
      this.y /= v.y;
    }
    return this;
  }

  limit(max) {
    if (this.magnitude() > max) {
      this.normalize();
      this.multiply(max);
    }
    return this;
  }

  magnitude() {
    return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2));
  }

  diff(v) {
    return new Vector(
      Math.abs(Math.abs(this.x) - Math.abs(v.x)),
      Math.abs(Math.abs(this.y) - Math.abs(v.y))
    );
  }

  normalize() {
    const magnitude = this.magnitude();
    if (magnitude !== 0) {
      this.divide(magnitude);
      this.x = Math.abs(this.x);
      this.y = Math.abs(this.y);
    }
    return this;
  }

  round() {
    this.x = Math.round(this.x);
    this.y = Math.round(this.y);
    return this;
  }

  render() {
    return Math.round(this.x) + ", " + Math.round(this.y);
  }
}

function _vectorOperation(v, op, arg) {
  const out = new Vector(v.x, v.y);
  out[op](arg);
  return out;
}
