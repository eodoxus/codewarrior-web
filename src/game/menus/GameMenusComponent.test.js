import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import GameMenus from "./GameMenusComponent";

import dialog from "../../../public/dialog.json";
import GameEvent from "../engine/GameEvent";
import Vector from "../engine/Vector";

const mockHero = {
  getPosition: () => new Vector(0, 0),
  getInventory: () => {
    return {
      get: () => {
        return {
          getSpells: () => []
        };
      }
    };
  }
};

beforeEach(() => {
  fetch.mockResponse(JSON.stringify(dialog));
});

describe("<GameMenus />", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(<GameMenus />, div);
    div.remove();
  });

  describe("Snapshots", () => {
    it("matches default snapshot", () => {
      const node = renderer.create(<GameMenus />);
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
      node.unmount();
    });

    it("has hero menu when hero menu is opened snapshot", () => {
      const node = renderer.create(<GameMenus />);
      GameEvent.fire(GameEvent.OPEN_HERO_MENU, mockHero);
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
      node.unmount();
    });

    it("has dialog when a dialog is active", () => {
      const node = renderer.create(<GameMenus />);
      GameEvent.fire(GameEvent.DIALOG, "test dialog");
      const tree = node.toJSON();
      expect(tree).toMatchSnapshot();
      node.unmount();
    });
  });
});
