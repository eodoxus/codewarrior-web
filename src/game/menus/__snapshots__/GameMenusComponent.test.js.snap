// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameMenus /> Snapshots has dialog when a dialog is active 1`] = `
<div
  className={undefined}
>
  
  
  <div
    className={undefined}
  >
    <div>
      test dialog
    </div>
  </div>
  <div
    className=""
  />
  
</div>
`;

exports[`<GameMenus /> Snapshots has hero menu when hero menu is opened snapshot 1`] = `
<div
  className={undefined}
>
  <div
    className={undefined}
    onClick={[Function]}
    style={
      Object {
        "left": 0,
        "top": 0,
      }
    }
  >
    <div
      className={undefined}
    >
      <div
        onClick={[Function]}
        style={
          Object {
            "background": "url(/sprites/hero-menu.png) no-repeat 0 0",
            "height": 16,
            "width": 16,
          }
        }
      />
    </div>
  </div>
  
  
  <div
    className=""
  />
  
</div>
`;

exports[`<GameMenus /> Snapshots matches default snapshot 1`] = `
<div
  className={undefined}
>
  
  
  
  <div
    className=""
  />
  
</div>
`;
